### PR TITLE
[WIP] Weapon cursor improvements

### PR DESCRIPTION
**Pull request by MadmanMartian**
Moves more cursor code to mob level

Rather than using 6 static sprites for recoil, the icon is generated procedurally. These icons are stored in an icon buffer for later use

Current issues:
 - ~~Cursor icon flickers when generating the sprite. This could be remedied by generating the sprites at game start, but would mean we would have sprites in buffer that could potentially never be used~~ The sprites are now generated at roundstart, and are streamed to the client on login
 - ~~Doesn't scale down like it used to.~~ recoil was being less than 10 (what the previous icon was at by default), changed the math

### CODE REVIEW

File: code/modules/mob/living/recoil.dm [hunk 1]
```diff
@@ -62,4 +62,54 @@
...
+				var/pixel_x
+				if(dir & NORTH)
+					pixel_y = CLAMP(offset, -MAX_ACCURACY_OFFSET, MAX_ACCURACY_OFFSET)
+				if(dir & SOUTH)
+					pixel_y = CLAMP(-offset, -MAX_ACCURACY_OFFSET, MAX_ACCURACY_OFFSET)
+				if(dir & EAST)
+					pixel_x = CLAMP(offset, -MAX_ACCURACY_OFFSET, MAX_ACCURACY_OFFSET)
+				if(dir & WEST)
+					pixel_x = CLAMP(-offset, -MAX_ACCURACY_OFFSET, MAX_ACCURACY_OFFSET)
+				to_chat(world, "overlay [dir] [overlay] [pixel_x] [pixel_y]")
```
Comment (MadmanMartian) [inline code/modules/mob/living/recoil.dm]: Debug, will remove in final

**Comment by SpadesSlick** [comment]
> Cursor icon flickers when generating the sprite. This could be remedied by generating the sprites at game start, but would mean we would have sprites in buffer that could potentially never be used

Should not be much of a problem, do that.

**Comment by mattroks101** [comment]
What do you mean by doesn't scale down like it used to? The visual doesn't look the same? Or it doesn't scale down at all?

**Comment by MadmanMartian** [comment]
It seems to stop before fully switching to scaled down

File: code/modules/mob/living/recoil.dm [hunk 2]
```diff
@@ -43,23 +37,57 @@
...
-			bottom = MAX_ACCURACY_OFFSET
-	if(bottom)
-		var/reduction = calc_reduction()
-		if(reduction > 0)
-			recoil_timer = addtimer(CALLBACK(src, .proc/update_recoil_cursor, G), 1 + (recoil - bottom) / reduction)
+/mob/living/proc/update_recoil_cursor()
+	update_cursor()
+	var/reduction = calc_reduction()
+	if(reduction > 0 && recoil > 0)
+		recoil_timer = addtimer(CALLBACK(src, .proc/update_recoil_cursor), 1+round(recoil/10)/reduction)
```
Comment (MadmanMartian) [inline code/modules/mob/living/recoil.dm]: Need to fix the math, reduction too slow compared to previous

1+round(recoil/10)/reduction

File: code/modules/mob/living/recoil.dm [hunk 3]
```diff
@@ -43,23 +42,66 @@
...
+		deltimer(recoil_timer)
+		recoil_timer = addtimer(CALLBACK(src, .proc/update_recoil_cursor), timer, TIMER_STOPPABLE)
+
+/mob/living/proc/update_cursor()
+	if(get_preference_value(/datum/client_preference/gun_cursor) != GLOB.PREF_YES)
+		remove_cursor()
+		return
+	if(client)
+		client.mouse_pointer_icon = initial(client.mouse_pointer_icon)
+		var/offset = round(calc_recoil())
```
Comment (PlinHost) [inline code/modules/mob/living/recoil.dm]:
```suggestion
		var/offset = min(round(calc_recoil()), MAX_ACCURACY_OFFSET)
```
Apparently recoil can be higher than `MAX_ACCURACY_OFFSET` and it causes next assert to fail.

File: code/modules/mob/living/recoil.dm [hunk 4]
```diff
@@ -43,23 +42,66 @@
...
+	var/list/L = GLOB.cursor_icons[icon_file]
+	if(L)
+		return L["[offset]"]
+
+/proc/add_cursor_icon(var/icon/icon, var/icon_file, var/offset)
+	var/list/L = GLOB.cursor_icons[icon_file]
+	if(!L)
+		GLOB.cursor_icons[icon_file] = list()
+		L = GLOB.cursor_icons[icon_file]
+	L.Add("[offset]")
```
Comment (PlinHost) [inline code/modules/mob/living/recoil.dm]: I am pretty sure this line is not needed.

**Comment by PlinHost** [comment]
Also reduction is a real mess, but I guess it is outside the scope of this PR.

File: code/modules/mob/living/recoil.dm (hunk 3 above)
Comment (MadmanMartian) [reply code/modules/mob/living/recoil.dm]: How were you able to do this?

Comment (PlinHost) [reply code/modules/mob/living/recoil.dm]: By shooting a full mag from Wintermute on full auto. Not that hard.

Comment (MadmanMartian) [reply code/modules/mob/living/recoil.dm]: What is the object code for that?

Comment (PlinHost) [reply code/modules/mob/living/recoil.dm]: I am not sure what code you are asking me for. Anyway, I have looked into the problem and found out that recoil should be capped by `MAX_ACCURACY_OFFSET`, but the capping code https://github.com/discordia-space/CEV-Eris/blob/57b2c11fe180717b7156be213c8cea197381225b/code/modules/projectiles/gun.dm#L349 was lost in #4313. It should be added back to `/mob/living/proc/handle_recoil`.

Comment (MadmanMartian) [reply code/modules/mob/living/recoil.dm]: The object path, sorry. Like /obj/item/weapon/projectile/lmg or something like that

Comment (PlinHost) [reply code/modules/mob/living/recoil.dm]: That is `/obj/item/weapon/gun/projectile/automatic/IH_heavyrifle`.

**Comment by MadmanMartian** [comment]
Yeah, only reason it looked smooth before was because of the timer stacking. Will continue to look into smoothing it out

**Comment by SpadesSlick** [comment]
It should work for now, we will improve it later.

**Pull request merged**